### ISSUE TITLE
Invalid Topology from GeometryPrecisionReducer with Polygon input

### ISSUE THREAD

**Issue by Sarev0k** [locationtech/jts#324]
When attempting to reduce the precision of the following polygon, I got some unexpected results.

### Attempted
Reduce the precision of the following polygon to six decimal places
`POLYGON ((-121.816901763 37.3285521, -121.817392418 37.328343575, -121.817876714 37.328137752, -121.818253966 37.327977421, -121.8184719 37.3278848, -121.816901763 37.3285521))`

### Expected Result
Polygon with reduced precision and a valid topology

### Actual Result
The following multipolygon which contains an invalid topology:
`POLYGON ((-121.817976 37.328096, -121.817877 37.328138, -121.817392 37.328344, -121.816902 37.328552, -121.817976 37.328096), (-121.817976 37.328096, -121.818472 37.327885, -121.818254 37.327977, -121.817976 37.328096))`

When using IsValidOp to get a more details I get the following explanation:
`Self-intersection at or near point (-121.81766914284646, 37.32822628571883, NaN)`

### Reproduction

```java
final WKTReader reader = new WKTReader();
final Geometry geometry =
    reader.read("POLYGON ((-121.816901763 37.3285521, -121.817392418 37.328343575, " +
                "-121.817876714 37.328137752, -121.818253966 37.327977421, " +
                "-121.8184719 37.3278848, -121.816901763 37.3285521))");
assertTrue(geometry.isValid()); // this passes

final Geometry reducedGeometry =
    GeometryPrecisionReducer.reduce(geometry, new PrecisionModel(Math.pow(10, 6)));

assertTrue(reducedGeometry.isValid()); // this fails
```

**Comment by dr-jts** [comment]
This is occurring because the `GeometryPrecisionReducer `actually uses a bit of a hack to attempt to create valid polygon topology.  It performs a `buffer(0)` of the reduced-precision input.  This works in many cases, but not all.  This is a case where it doesn't work, unfortunately.

I did notice that if you run `reduce `again, the output is then valid.  But this probably isn't a general solution.

**Comment by dr-jts** [comment]
Some further analysis of this failure case:  It's actually pretty pathological, because the "width" of the polygon is much less that the precision it is being rounded to.  (The width can be determined approximately by using `[MinimumClearance.getDistance](http://locationtech.github.io/jts/javadoc/org/locationtech/jts/precision/MinimumClearance.html#getDistance())` , which shows a separation of only about 1e-10.)  This means that when the vertices are rounded, there are inevitably self-intersections introduced.   And in fact there is a "second-order" self-intersection, meaning that when the initial intersection is noded (in the required precision model) there is a further self-intersection created.  This self-intersection appears in the shell of the polygon produced by `buffer(0)`.  The result is that the output polygon is invalid.  Also, because the shell is invalid the buffer algorithm incorrectly creates an "outside hole".

Handling this kind of situation is totally outside the design parameters of the buffer algorithm, so it's not surprising that it doesn't produce desirable output.

As a further note, at the moment my best idea for implementing a foolproof precision reduction algorithm is to use snap-rounding to produce a guaranteed valid topology.  But note that snap-rounding is a more aggressive noding approach, which in this case would reduce the polygon to a linestring.  This is a reasonable output, given the size of the precision model versus the width of the polygon.

**Comment by dr-jts** [comment]
A sidenote: this case also illustrates the difficulty in implementing robust noding.  When node vertices are introduced into a linear arrangement, due to rounding they can perturb the arrangement so far that further segment intersections are introduced.  (This is most obvious when using a defined precision model, but it also happens when using double-precision floating point).  This situation can repeat, and in pathological cases can require a very large number of iterations to compute a correct noding.

The benefit of snap-rounding is that it avoids this issue, by guaranteeing that vertices are far enough apart that segments cannot move far enough to introduce new intersections.  The downside is that the input vertices must be rounded to the precision grid ahead of time.

**Comment by dr-jts** [comment]
Fixed by #648
